src/utils.c

```diff
@@ -339,6 +340,109 @@ timestamptz_bucket(PG_FUNCTION_ARGS)
 	PG_RETURN_TIMESTAMPTZ(result);
 }
 
+TS_FUNCTION_INFO_V1(ts_int16_bucket);
+Datum
+ts_int16_bucket(PG_FUNCTION_ARGS)
+{
+	int16		period = PG_GETARG_INT16(0);
+	int16		timestamp = PG_GETARG_INT16(1);
+	int16		result;
+
+	result = (timestamp / period) * period;
+	if (timestamp < 0)
+	{
+		/*
+		 * If there is a remainder we need to subtract another period since
+		 * division truncates toward 0 in C99.
+		 */
+		if (timestamp % period)
+		{
+			/*
+			 * Set result to INT16_MIN on integer underflow.
+			 */
+			if (result < INT16_MIN + period)
+			{
+				result = INT16_MIN;
+			}
+			else
+			{
+				result = result - period;
+			}
+
+		}
+	}
+	PG_RETURN_INT16(result);
+}
+
+TS_FUNCTION_INFO_V1(ts_int32_bucket);
+Datum
+ts_int32_bucket(PG_FUNCTION_ARGS)
+{
+	int32		period = PG_GETARG_INT32(0);
+	int32		timestamp = PG_GETARG_INT32(1);
+	int32		result;
+
+	result = (timestamp / period) * period;
+	if (timestamp < 0)
+	{
+		/*
+		 * If there is a remainder we need to subtract another period since
+		 * division truncates toward 0 in C99.
+		 */
+		if (timestamp % period)
+		{
+			/*
+			 * Set result to INT32_MIN on integer underflow.
+			 */
+			if (result < INT32_MIN + period)
+			{
+				result = INT32_MIN;
+			}
+			else
+			{
+				result = result - period;
+			}
+
+		}
+	}
+	PG_RETURN_INT32(result);
+}
+
+
+TS_FUNCTION_INFO_V1(ts_int64_bucket);
+Datum
+ts_int64_bucket(PG_FUNCTION_ARGS)
+{
+	int64		period = PG_GETARG_INT64(0);
+	int64		timestamp = PG_GETARG_INT64(1);
+	int64		result;
+
+	result = (timestamp / period) * period;
```

Comment:
On `x86_64` this line will crash if timestamp is `INT64_MIN` and `period` is `-1`, as will the 32 bit version.

Reply:
Hmm do we want to catch that? The assumption is that period > 0, pretty sure other weird stuff will crop up with negative period. We would have to check for period being non-zero aswell. Dont think the non-int time_bucket functions check for > 0 interval.